packages/react-dom/src/client/focus/TabFocusController.js

```diff
@@ -0,0 +1,180 @@
+      const firstTabbableElem = tabbableNodes[0];
+      const lastTabbableElem = tabbableNodes[tabbableNodes.length - 1];
+      const currentIndex = tabbableNodes.indexOf(document.activeElement);
+      let focusedElement = null;
+      if (currentIndex !== -1) {
+        focusedElement = tabbableNodes[currentIndex];
+      }
+      return [
+        tabbableNodes,
+        firstTabbableElem,
+        lastTabbableElem,
+        currentIndex,
+        focusedElement,
+      ];
+    }
+
+    function focusFirst(): void {
+      const [, firstTabbableElem] = getTabbableNodes();
+      firstTabbableElem.focus();
+    }
+
+    function focusNext(): boolean {
+      const [
+        tabbableNodes,
+        firstTabbableElem,
+        lastTabbableElem,
+        currentIndex,
+        focusedElement,
+      ] = getTabbableNodes();
+
+      if (focusedElement === null) {
+        firstTabbableElem.focus();
+      } else if (focusedElement === lastTabbableElem) {
+        if (contain === true) {
+          firstTabbableElem.focus();
+        } else {
+          return true;
+        }
+      } else {
+        tabbableNodes[currentIndex + 1].focus();
+      }
+      return false;
+    }
+
+    function focusPrevious(): boolean {
+      const [
+        tabbableNodes,
+        firstTabbableElem,
+        lastTabbableElem,
+        currentIndex,
+        focusedElement,
+      ] = getTabbableNodes();
+
+      if (focusedElement === null) {
+        firstTabbableElem.focus();
+      } else if (focusedElement === firstTabbableElem) {
+        if (contain === true) {
+          lastTabbableElem.focus();
+        } else {
+          return true;
+        }
+      } else {
+        tabbableNodes[currentIndex - 1].focus();
+      }
+      return false;
+    }
+
+    function getPreviousController(): null | ControllerHandle {
+      const tabbableScope = scopeRef.current;
+      const allScopes = tabbableScope.getChildrenFromRoot();
+      if (allScopes === null) {
+        return null;
+      }
+      const currentScopeIndex = allScopes.indexOf(tabbableScope);
+      if (currentScopeIndex === -1 || currentScopeIndex === 0) {
+        return null;
+      }
+      return allScopes[currentScopeIndex - 1].getHandle();
+    }
+
+    function getNextController(): null | ControllerHandle {
+      const tabbableScope = scopeRef.current;
+      const allScopes = tabbableScope.getChildrenFromRoot();
+      if (allScopes === null) {
+        return null;
+      }
+      const currentScopeIndex = allScopes.indexOf(tabbableScope);
+      if (
+        currentScopeIndex === -1 ||
+        currentScopeIndex === allScopes.length - 1
+      ) {
+        return null;
+      }
+      return allScopes[currentScopeIndex + 1].getHandle();
+    }
+
+    const controllerHandle: ControllerHandle = {
+      focusFirst,
+      focusNext,
+      focusPrevious,
+      getNextController,
+      getPreviousController,
+    };
+
+    useImperativeHandle(ref, () => controllerHandle);
```

Comment:
Note this hook is not good for perf according to my tests

Reply:
No other way to do what we want here, other than using a class. :/